sql-plugin/src/main/scala/com/nvidia/spark/rapids/PrioritySemaphore.scala

```diff
@@ -58,10 +58,21 @@ class PrioritySemaphore[T](val maxPermits: Int)(implicit ordering: Ordering[T])
     lock.lock()
     try {
       val condition = lock.newCondition()
+      var queued = false
       while (!canAcquire(numPermits)) {
-        waitingQueue.enqueue(ThreadInfo(priority, condition))
+        if (!queued) {
+          waitingQueue.enqueue(ThreadInfo(priority, condition, numPermits))
+          queued = true
+        }
         condition.await()
       }
+      if (queued) {
+        if (waitingQueue.isEmpty || waitingQueue.head.condition != condition) {
+          throw new IllegalStateException("acquire thought it queued but acquired semaphore "
+            + "without being at the head of the queue")
+        }
+        waitingQueue.dequeue()
```

Comment:
I think leaving the entry in the queue is going to be problematic.  There's the multi-wakeup mentioned elsewhere, along with the fact that if two threads currently own the semaphore, and both release, we don't want them both trying to wake up the same thread.  If the thread is taking a long time to wakeup for some reason, we shouldn't block other threads behind it.  I think we should dequeue as part of signaling as was done before.

Reply:
I agree with the point you made about priority inversion (in the other comment), but, as for the other concerns, my understanding is the `lock` should not allow multiple threads to `release` at the same time (similar to if it were a `synchronized` block instead). This would mean that two threads should not be able to wake the same thread, and also that a slow wakeup blocking another wakeup is unavoidable.

Reply:
> This would mean that two threads should not be able to wake the same thread

If the node is left in the queue, it very much will do so.  The second release, which will execute when the first release lets go of the lock, will find the same thread at the front of the queue and try to wake that up.

Reply:
Oh I see what you mean, the waiting thread is not guaranteed to wake up and resume execution before the second release proceeds.